### ISSUE TITLE
[core] PMD should warn about multiple instances of the same rule in a ruleset

### ISSUE THREAD

**Issue by Fernal73** [pmd/pmd#2006]
**Affects PMD Version:** 
6.17.0

**Rule:**
All rules.

**Description:**
PMD does not flag or display a warning if the ruleset contains more than one instance of a rule.

https://pmd.github.io/latest/pmd_userdocs_making_rulesets.html

**Code Sample demonstrating the issue:**

Referring : https://github.com/pmd/pmd/issues/1904#event-2615431496

```xml
<rule ref="category/java/errorprone.xml/EmptyCatchBlock">
    <properties>
        <property name="allowCommentedBlocks" value="true" />
    </properties>
</rule>
<rule ref="category/java/errorprone.xml/EmptyCatchBlock"/>
```

**Running PMD through:**
[CLI | Ant | Maven | Gradle | Designer | Other]

**Expected:**

A warning must be displayed if multiple instances of a rule are (permissibly) configured.

**Comment by andipabst** [comment]
The method [Ruleset#addRuleReplaceIfExists](https://github.com/pmd/pmd/blob/cfb4fbec0ff902278e2502b79b8aa0b87d347c74/pmd-core/src/main/java/net/sourceforge/pmd/RuleSet.java#L164) silently overwrites duplicate rules in a ruleset. The Javadoc for this method states

> If a rule with the same name and language already existed before in the ruleset, then the new rule will replace it. This makes sure that the rule configured is overridden.

I have added a Log output to the method and created a pull request  (#2059).

**Comment by Fernal73** [comment]
I noticed that while ignoring rules that don't meet the priority criteria,
no message is displayed to the console while not in debug mode. It would be
a good idea to state that so many rules were skipped or ignored and to run
in debug mode if you wanted to check which or reset the minimum priority
level. I changed my minimum priority to 1 and received no such warnings in
the console.

**Comment by Fernal73** [comment]
It might also be a good idea to label the priority levels such as INFO,
WARNING, HIGH, VERY HIGH and SEVERE instead. Using flag minimumPriority and
setting it to high number 5 is not really intuitive.

**Comment by Fernal73** [comment]
Rules ignored i.e., their count must be logged to console.

**Comment by Fernal73** [comment]
I wouldn't be averse to adding another switch warnomitted or warnskipped if
the console feels too verbose with this additional warning. Better still,
use a properties file like the ant task does for these varied options. It's
extensible as well.

**Comment by Fernal73** [comment]
Correction: The ant task doesn't use a properties file. I wish it did,
though.

**Comment by adangel** [comment]
I see the following test cases, that we need to cover:

## Duplicated rule reference
See also #2059 
Ruleset:
```
<?xml version="1.0"?>

<ruleset name="Custom Rules"
    xmlns="http://pmd.sourceforge.net/ruleset/2.0.0"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://pmd.sourceforge.net/ruleset/2.0.0 https://pmd.sourceforge.io/ruleset_2_0_0.xsd">

    <description>My custom rules</description>

    <rule ref="category/java/errorprone.xml/InvalidLogMessageFormat" />
    <rule ref="category/java/errorprone.xml/InvalidLogMessageFormat" />
</ruleset>
```
Expected: Log message about duplicated rule in the ruleset

## Duplicated rule reference with override
Ruleset:
```
<?xml version="1.0"?>

<ruleset name="Custom Rules"
    xmlns="http://pmd.sourceforge.net/ruleset/2.0.0"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://pmd.sourceforge.net/ruleset/2.0.0 https://pmd.sourceforge.io/ruleset_2_0_0.xsd">

    <description>My custom rules</description>

    <rule ref="category/java/errorprone.xml" />
    <rule ref="category/java/errorprone.xml/InvalidLogMessageFormat">
        <priority>1</priority>
    </rule>
</ruleset>
```
Expected: This should not produce a log message, since e.g. the priority is overridden.

## Duplicated rule reference with unintended override
Ruleset:
```
<?xml version="1.0"?>

<ruleset name="Custom Rules"
    xmlns="http://pmd.sourceforge.net/ruleset/2.0.0"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://pmd.sourceforge.net/ruleset/2.0.0 https://pmd.sourceforge.io/ruleset_2_0_0.xsd">

    <description>My custom rules</description>

    <rule ref="category/java/errorprone.xml/InvalidLogMessageFormat">
        <priority>1</priority>
    </rule>
    <rule ref="category/java/errorprone.xml" />
</ruleset>
```
Expected: This however should produce a warning, since with referencing the complete category, the single reference rule is overridden again, so the priority is ignored.

## Duplicated rule references with bulk references
Ruleset:
```
<?xml version="1.0"?>

<ruleset name="Custom Rules"
    xmlns="http://pmd.sourceforge.net/ruleset/2.0.0"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://pmd.sourceforge.net/ruleset/2.0.0 https://pmd.sourceforge.io/ruleset_2_0_0.xsd">

    <description>My custom rules</description>

    <rule ref="category/java/errorprone.xml" />
    <rule ref="category/java/errorprone.xml" />

    <rule ref="category/java/errorprone.xml/InvalidLogMessageFormat">
        <priority>1</priority>
    </rule>
    <rule ref="category/java/errorprone.xml/InvalidLogMessageFormat">
        <priority>2</priority>
    </rule>
</ruleset>
```
Expected: This should produce two warnings:
* one for the second bulk reference of errorprone
* one for the second rule reference of InvalidLogMessageFormat

**Comment by andipabst** [comment]
Okay, thank you for these examples, I will implement them over the next couple of days and then update the pull request.

**Comment by Fernal73** [comment]
@adangel 
I haven't used bulk references in my configuration but wouldn't you expect duplication if a rule from that category is added without excepting it in the bulk reference? Is PMD handling that scenario currently? 

https://pmd.github.io/latest/pmd_userdocs_making_rulesets.html#bulk-adding-rules

**Comment by adangel** [comment]
> Expected: This should produce one warning:
> for the second rule instance InvalidLogMessageFormat

That's not how we see that it should be working. If you first bulk import a complete ruleset/category and then refine a single rule, that should be ok. If you write this in your ruleset, we assume, you want exactly what you have written. This enables simple overriding.
The behavior here is of course debatable, but I think,  this pattern could be more widespread than we imagine. Fact is, that it currently works without a warning, so that's the main reason to allow this ruleset "pattern" (compatibility...).
I see however, that we don't explicitly describe in the documentation, that this "feature" will actually work.

Asked in a different way: What would you expect from PMD to do, if you would need to execute such a ruleset? Fail, because InvalidLogMessage is twice defined, run InvalidLogMessage twice, because it is defined twice, ...

**Comment by Fernal73** [comment]
If that's the existing behaviour, there's no need to change it but document it. If you don't expect too many of these rules to be overridden, there might be a case to add a child element 'override' to explicitly model this behaviour (similar to 'except').

**Comment by oowekyala** [comment]
> Do we foresee that PMD will need to allow multiple configurations of the same rule i.e., multiple instances that coexist? Checkstyle handles this by allowing each instance to be identified by a user-specified element 'id'.

AFAIK if you rename a rule reference then it is not equal to the old one anymore and both coexist. This is an important element in eg what's discussed [here](https://github.com/pmd/pmd/wiki/Property-framework-7-0-0#use-case) for 7.0, because it allows making more generic rules.

**Comment by Fernal73** [comment]
The express purpose of renaming the rules (in Checkstyle) was to identify the source of the violation in the output. The rules engine doesn't care if there are multiple instances, unlike PMD which utilizes the last one defined. 
IMO, generic rules are meant for more advanced users i.e., power users. 
There will still be a need for grouping of rules for novices.

**Comment by oowekyala** [comment]
> IMO, generic rules are meant for more advanced users i.e., power users.

Maybe, though honestly the concept is pretty simple (boils down to *some properties are required for the rule to run, and have no default value*) and it should be pretty easy to explain. We'll see about that when preparing this feature for 7.0.

> There will still be a need for grouping of rules for novices.

Do you mean rulesets? I don't see the connection to rule renaming.

> Is a tutorial needed in the documentation about starting out with PMD and getting beginners started? There can be lessons added later how to go about performing more advanced tasks.

Yeah I had written up a new tutorial but it's already outdated as the designer has changed a lot since then (6.10.0). I mean it's less outdated than the current doc, but still, not mergeable. For reference, it's in [this branch](https://github.com/oowekyala/pmd/tree/docs-rules-newer/docs/pages/pmd/userdocs/extending), pages `writing_.*_rules.md`

**Comment by Fernal73** [comment]
My concern with rulesets, generic rules and renaming rules is that rulesets
may contain two or more instances of generic rules.

The possibility appears remote and is probably a theoretical exercise on my
part.

Then you'd have a generic rule, a renamed rule or more and these may or may
not be part of the same ruleset. Placing them in rule sets will not make
them visible to the user as rooted in the same generic rule. Using them
individually will.

That's the nub of the matter.

Checkstyle doesn't have rulesets but rules are categorised in their
documentation.

Not having rulesets will make managing the codebase simpler and easier.

How will it impact usability?

**Comment by Fernal73** [comment]
It's my understanding , now , that you can simply take the pre-defined rulesets and create a six to 10 line rules document, depending on which rulesets you need. That can be a good starting point.

**Comment by oowekyala** [comment]
> My concern with rulesets, generic rules and renaming rules is that rulesets
may contain two or more instances of generic rules.

This is the whole point of "generic rules". We can ship a single rule, and users may derive any number of concrete rules from it, by changing the configuration. 



>  Placing them in rule sets will not make them visible to the user as rooted in the same generic rule.

This is a feature. "Generic rules" are a way for rules to share an *implementation*. Users are not supposed to care how rules are implemented, hence it's better off hidden anyway.

Besides, I really don't follow where you're going. If I understand correctly, at some point in this conversation you seem to have shifted your stance to advocating for *removing rulesets*? If so this looks off-topic and I'd ask you open a separate issue to make your case

> It's my understanding , now , that you can simply take the pre-defined rulesets and create a six to 10 line rules document, depending on which rulesets you need. That can be a good starting point.

The recommended starting point is the `quickstart.xml` ruleset. If you want to make your own ruleset, use [this tutorial](https://pmd.github.io/latest/pmd_userdocs_making_rulesets.html). You can even include `quickstart.xml` to bootstrap your ruleset.

**Comment by Fernal73** [comment]
- "The currently ~60 classes framework can be implemented with just ~10
   classes"
   - "Blank properties allow more abstraction in the rules we ship,
   allowing us to ship and maintain less rules"



Maybe, I misunderstood the above two points.

https://github.com/pmd/pmd/wiki/Property-framework-7-0-0#use-case

As for creating rules and rulesets, that's not how I created my set of PMD
rules that I use. Frankly, I can't even recall how I did it.

I do recall starting out with the installation of PMD that ran with the
shell script, only to switch over at some point to using the download from
Maven repository.

https://github.com/Fernal73/LearnJava/blob/master/pmdrules.xml

I notice that this actually has the rule set named quickstart.

I guess what I'm referring to is the ability to included entire sets of
rules as:

<rule ref="category/jsp/design.xml"/>

Am I mistaken in assuming that this
runs all the rules under this category?
